src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.SymbolSet.cs

```diff
@@ -0,0 +1,268 @@
+                {
+                    var currentSymbol = workQueue.Pop();
+                    await AddUpSymbolsAsync(engine, currentSymbol, upSymbols, workQueue, allProjects, cancellationToken).ConfigureAwait(false);
+                }
+
+                return upSymbols;
+            }
+
+            protected static async Task AddCascadedAndLinkedSymbolsToAsync(
+                FindReferencesSearchEngine engine, ImmutableArray<ISymbol> symbols, HashSet<ISymbol> to, Stack<ISymbol> workQueue, CancellationToken cancellationToken)
+            {
+                foreach (var symbol in symbols)
+                    await AddCascadedAndLinkedSymbolsToAsync(engine, symbol, to, workQueue, cancellationToken).ConfigureAwait(false);
+            }
+
+            protected static async Task AddCascadedAndLinkedSymbolsToAsync(
+                FindReferencesSearchEngine engine, ISymbol symbol, HashSet<ISymbol> to, Stack<ISymbol> workQueue, CancellationToken cancellationToken)
+            {
+                var solution = engine._solution;
+                symbol = await MapToAppropriateSymbolAsync(solution, symbol, cancellationToken).ConfigureAwait(false);
+
+                foreach (var finder in engine._finders)
+                {
+                    var cascaded = await finder.DetermineCascadedSymbolsAsync(symbol, solution, engine._options, cancellationToken).ConfigureAwait(false);
+                    foreach (var cascade in cascaded)
+                        await AddLinkedSymbolsAsync(cascade).ConfigureAwait(false);
+                }
+
+                await AddLinkedSymbolsAsync(symbol).ConfigureAwait(false);
+
+                return;
+
+                async Task AddLinkedSymbolsAsync(ISymbol symbol)
+                {
+                    symbol = await MapToAppropriateSymbolAsync(solution, symbol, cancellationToken).ConfigureAwait(false);
+
+                    foreach (var linked in await SymbolFinder.FindLinkedSymbolsAsync(symbol, solution, cancellationToken).ConfigureAwait(false))
+                    {
+                        if (to.Add(linked))
+                            workQueue.Push(linked);
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Finds all the symbols 'down' the inheritance hierarchy of <paramref name="symbol"/> in the given
+            /// project.  The symbols found are added to <paramref name="to"/>.  If <paramref name="to"/> did not
+            /// contain that symbol, then it is also added to <paramref name="workQueue"/> to allow fixed point
+            /// algorithms to continue.
+            /// </summary>
+            /// <remarks><paramref name="projects"/> will always be a single project.  We just pass this in as a set to
+            /// avoid allocating a fresh set every time this calls into FindMemberImplementationsArrayAsync.
+            /// </remarks>
+            protected async Task AddDownSymbolsAsync(
```

Comment:
In AddUpSymbolsAsync, the engine is passed in and is static (I believe because it is being used in CreateAsync), but here it uses this.Engine.  Maybe pass in engine here just for consistency?  I was trying to figure out earlier why BidirectionalSymbolSet passed in the engine separately for only one of the requests.

Reply:
Gotcha.  Yeah, it's because we always start with the initial and 'up' set when creating these sets (so it is computed prior to having an instance). But then we compute the latter values in the instance. 

However, I could move the initial set computation into the set, which could potential resolve this. I'll see what I can do.

Reply:
unified this.